### PR TITLE
[REEF-1992] Upgrade Azure Storage from 6.1 to 8.1.3

### PR DESCRIPTION
**Pull request by tyclintw**
[REEF-1992] Upgrade Azure Storage from 6.1 to 8.1.3

    Upgrades all Azure Storage dependencies to version 8.1.3. Change is necessary for external consumers of REEF as well as matching the Org.Apache.Reef.DotNet.sln.
   Minor function call changes to address API differences. Update tests to address functional differences.
  
JIRA:
  [REEF-1992](https://issues.apache.org/jira/browse/REEF-1992)

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj [hunk 1]
```diff
@@ -44,16 +44,16 @@ under the License.
...
     </Reference>
-    <Reference Include="Microsoft.Data.OData, Version=5.6.4.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>$(PackagesDir)\Microsoft.Data.OData.5.6.4\lib\net40\Microsoft.Data.OData.dll</HintPath>
+    <Reference Include="Microsoft.Data.OData, Version=5.8.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>$(PackagesDir)\Microsoft.Data.OData.5.8.2\lib\net40\Microsoft.Data.OData.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Data.Services.Client, Version=5.6.4.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>$(PackagesDir)\Microsoft.Data.Services.Client.5.6.4\lib\net40\Microsoft.Data.Services.Client.dll</HintPath>
+    <Reference Include="Microsoft.Data.Services.Client, Version=5.8.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (singlis) [inline lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj]: Why is the Version 5.8.1.0, but the hint path is 5.8.2?

File: lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj [hunk 2]
```diff
@@ -88,8 +88,8 @@ under the License.
     <Reference Include="System.Core" />
     <Reference Include="System.Management" />
     <Reference Include="System.Net.Http" />
-    <Reference Include="System.Spatial, Version=5.6.4.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>$(PackagesDir)\System.Spatial.5.6.4\lib\net40\System.Spatial.dll</HintPath>
+    <Reference Include="System.Spatial, Version=5.8.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (singlis) [inline lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj]: Path is different here as well.

File: lang/cs/Org.Apache.REEF.Tests/Org.Apache.REEF.Tests.csproj [hunk 3]
```diff
@@ -38,20 +38,20 @@ under the License.
       <HintPath>$(PackagesDir)\Microsoft.Azure.KeyVault.Core.1.0.0\lib\net40\Microsoft.Azure.KeyVault.Core.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Data.Edm, Version=5.6.4.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>$(PackagesDir)\Microsoft.Data.Edm.5.6.4\lib\net40\Microsoft.Data.Edm.dll</HintPath>
+    <Reference Include="Microsoft.Data.Edm, Version=5.8.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (singlis) [inline lang/cs/Org.Apache.REEF.Tests/Org.Apache.REEF.Tests.csproj]: In this file as well...

File: lang/cs/Org.Apache.REEF.IO.Tests/Org.Apache.REEF.IO.Tests.csproj [hunk 4]
```diff
@@ -124,6 +138,9 @@ under the License.
       <SubType>Designer</SubType>
     </None>
   </ItemGroup>
+  <ItemGroup>
```
Comment (singlis) [inline lang/cs/Org.Apache.REEF.IO.Tests/Org.Apache.REEF.IO.Tests.csproj]: Is this change intentional?

File: lang/cs/Org.Apache.REEF.IO.Tests/TestAzureBlockBlobFileSystem.cs [hunk 5]
```diff
@@ -18,7 +18,6 @@
 using System;
 using System.IO;
 using System.Linq;
-using System.Threading.Tasks;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO.Tests/TestAzureBlockBlobFileSystem.cs]: This seems like an unrelated change.

File: lang/cs/Org.Apache.REEF.IO.Tests/TestAzureBlockBlobFileSystemE2E.cs [hunk 6]
```diff
@@ -169,6 +169,8 @@ public void TestCopyToLocalE2E()
             var helloFilePath = PathToFile(HelloFile);
             var blob = _container.GetBlockBlobReference(HelloFile);
             var tempFilePath = Path.GetTempFileName();
+            File.Delete(tempFilePath);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO.Tests/TestAzureBlockBlobFileSystemE2E.cs]: Please document this change in the commit message or move it to another commit.

File: lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj (hunk 1 above)
Comment (tyclintw) [reply lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj]: These paths were auto generated by Visual Studio. Upon investigation, it appears fairly common for minor version number to not match.

File: lang/cs/Org.Apache.REEF.IO.Tests/TestAzureBlockBlobFileSystemE2E.cs (hunk 6 above)
Comment (tyclintw) [reply lang/cs/Org.Apache.REEF.IO.Tests/TestAzureBlockBlobFileSystemE2E.cs]: Message updated.

**Comment by singlis** [comment]
@tyclintw - the DotNet proj migration PR went through to master. Can you take a look at the .DotNet projects to determine if the library project changes are needed? 

Just looking at the build.DotNet.props and the csproj, it looks like we are using the same versions:

WindowAzureStorage - 8.1.3 (you have 8.1.2)
System.Spatial - 5.8.2
Microsoft.Data.Services.Client - 5.8.2
Microsoft.Data.OData - 5.8.2

File: lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj (hunk 2 above)
Comment (tyclintw) [reply lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj]: These paths were auto generated by Visual Studio. Upon investigation, it appears fairly common for minor version numbers to not match.

File: lang/cs/Org.Apache.REEF.Tests/Org.Apache.REEF.Tests.csproj (hunk 3 above)
Comment (tyclintw) [reply lang/cs/Org.Apache.REEF.Tests/Org.Apache.REEF.Tests.csproj]: These paths were auto generated by Visual Studio. Upon investigation, it appears fairly common for minor version numbers to not match.

**Comment by markusweimer** [comment]
@tyclintw , @singlis What's the status of this PR? Updated for another round of reviews?

File: lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj (hunk 2 above)
Comment (singlis) [reply lang/cs/Org.Apache.REEF.IO/Org.Apache.REEF.IO.csproj]: ok disregard my comment then, thanks for looking.

File: lang/cs/Org.Apache.REEF.IO/FileSystem/AzureBlob/AzureCloudBlockBlob.cs [hunk 7]
```diff
@@ -112,11 +112,7 @@ public void DownloadToFile(string path, FileMode mode)
 
         public void UploadFromFile(string path, FileMode mode)
         {
-            #if REEF_DOTNET_BUILD
```
Comment (singlis) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/AzureBlob/AzureCloudBlockBlob.cs]: So the WindowsStorage library is not consistent on the APIs that are provided between .net45 and .netstandard. Please use the Async version as this is available in both libraries.

**Comment by singlis** [comment]
Thanks for the ping, I just responded.

File: lang/cs/Org.Apache.REEF.IO/FileSystem/AzureBlob/AzureCloudBlockBlob.cs (hunk 7 above)
Comment (tyclintw) [reply lang/cs/Org.Apache.REEF.IO/FileSystem/AzureBlob/AzureCloudBlockBlob.cs]: Fixed to use the Async version.

**Comment by tyclintw** [comment]
@markusweimer All pending changes are in and ready for final review .

**Comment by tyclintw** [comment]
@markusweimer The .DotNet.csproj was already on storage 8.1.3 so no change is needed. This change brings the two projects inline with each other.

**Comment by markusweimer** [comment]
@tyclintw sorry to do this, but there are merge conflicts. Can you please resolve them?

**Comment by tyclintw** [comment]
@markusweimer Conflicts have been resolved.

**Comment by markusweimer** [comment]
I had to do some surgery during the merge. Hence, this commit looks as if it is from me. Sorry, @tyclintw !
